src/engine/renderer/tr_shader.cpp

```diff
@@ -5170,20 +5170,31 @@ static void CollapseStages()
 		stage->shaderHasNoLight = shaderHasNoLight;
 
 		// Available textures.
-		stage->hasNormalMap = stage->bundle[ TB_NORMALMAP ].image[ 0 ] != nullptr;
-		stage->hasHeightMap = stage->bundle[ TB_HEIGHTMAP ].image[ 0 ] != nullptr;
-		stage->isHeightMapInNormalMap = stage->isHeightMapInNormalMap && stage->hasNormalMap;
-		stage->hasMaterialMap = stage->bundle[ TB_MATERIALMAP ].image[ 0 ] != nullptr;
-		stage->hasGlowMap = stage->bundle[ TB_GLOWMAP ].image[ 0 ] != nullptr;
-		stage->isMaterialPhysical = stage->collapseType == collapseType_t::COLLAPSE_PBR;
+		bool hasNormalMap = stage->bundle[ TB_NORMALMAP ].image[ 0 ] != nullptr;
+		bool hasHeightMap = stage->bundle[ TB_HEIGHTMAP ].image[ 0 ] != nullptr;
+		bool hasMaterialMap = stage->bundle[ TB_MATERIALMAP ].image[ 0 ] != nullptr;
+		bool hasGlowMap = stage->bundle[ TB_GLOWMAP ].image[ 0 ] != nullptr;
+
+		// Texture storage variants.
+		stage->hasHeightMapInNormalMap &= hasNormalMap;
 
 		// Available features.
-		stage->enableNormalMapping = r_normalMapping->integer && stage->hasNormalMap;
-		stage->enableDeluxeMapping = r_deluxeMapping->integer && stage->hasNormalMap;
-		stage->enableReliefMapping = r_reliefMapping->integer && !shader.disableReliefMapping && ( stage->hasHeightMap || stage->isHeightMapInNormalMap );
-		stage->enablePhysicalMapping = r_physicalMapping->integer && stage->hasMaterialMap && stage->isMaterialPhysical;
-		stage->enableSpecularMapping = r_specularMapping->integer && stage->hasMaterialMap && !stage->isMaterialPhysical;
-		stage->enableGlowMapping = r_glowMapping->integer && stage->hasGlowMap;
+		stage->enableNormalMapping = hasNormalMap && r_normalMapping->integer;
+		stage->enableDeluxeMapping = hasNormalMap && r_deluxeMapping->integer;
+
+		stage->enableReliefMapping = ( hasHeightMap || stage->hasHeightMapInNormalMap )
+			&& !shader.disableReliefMapping && r_reliefMapping->integer;
+
+		stage->enableGlowMapping = hasGlowMap && r_glowMapping->integer;
+
+		if ( stage->collapseType == collapseType_t::COLLAPSE_PBR )
+		{
+			stage->enablePhysicalMapping = hasMaterialMap && r_physicalMapping->integer;
+		}
+		else
+		{
+			stage->enableSpecularMapping = hasMaterialMap && r_specularMapping->integer;
```

Comment:
Explicitly set the other to false?

ie,

```

stage->enablePhysicalMapping = hasMaterialMap && r_physicalMapping->integer;
stage->enableSpecularMapping = false;
```

Reply:
Done.